### PR TITLE
Minor Readability changes

### PR DESCRIPTION
**Pull request by carl-mastrangelo**

### CODE REVIEW

**Comment by carl-mastrangelo** [comment]
R= @ejona86

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java [hunk 1]
```diff
@@ -125,8 +130,8 @@ protected void inboundHeadersReceived(Metadata.Headers headers) {
    *
    * @param frame the received data frame. Its ownership is transferred to this method.
    */
-  protected void inboundDataReceived(ReadableBuffer frame) {
-    Preconditions.checkNotNull(frame, "frame");
```
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Why remove the "frame" string?

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java [hunk 2]
```diff
@@ -166,12 +172,15 @@ protected final void deframeFailed(Throwable cause) {
...
-    Preconditions.checkNotNull(trailers, "trailers");
+  protected final void inboundTrailersReceived(Metadata.Trailers trailers, Status status) {
+    checkNotNull(status);
+    checkNotNull(trailers);
     if (inboundPhase() == Phase.STATUS) {
-      log.log(Level.INFO, "Received trailers on closed stream {0}\n {1}\n {3}",
+      log.log(Level.INFO, "Received trailers on closed stream {0}\n {1}\n {2}",
           new Object[]{id(), status, trailers});
     }
+    checkState(status == null, "status has already been set");
```
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: This doesn't look right. It seems guaranteed to fail. Shouldn't it be `this.status`? Ditto on the next line.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java [hunk 3]
```diff
@@ -214,7 +223,7 @@ protected final void internalSendFrame(WritableBuffer frame, boolean endOfStream
    */
   public void transportReportStatus(final Status newStatus, boolean stopDelivery,
       final Metadata.Trailers trailers) {
-    Preconditions.checkNotNull(newStatus, "newStatus");
```
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Why remove "newStatus"?

File: core/src/main/java/io/grpc/transport/Http2ClientStream.java [hunk 4]
```diff
@@ -113,30 +114,31 @@ protected void transportHeadersReceived(Metadata.Headers headers) {
...
+    if (transportError == null) {
       inboundDataReceived(frame);
       if (endOfStream) {
         // This is a protocol violation as we expect to receive trailers.
         transportError = Status.INTERNAL.withDescription("Recevied EOS on DATA frame");
         inboundTransportError(transportError);
       }
+      return;
+    }
+    // We've already detected a transport error and now we're just accumulating more detail
```
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/Http2ClientStream.java]: This code is actually the error case. It seems strange to have it be the core code flow.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 1 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Redundant.  A stack trace will point to exactly what line caused the problem.  Additionally, the string frame is a minor maintainability drag, since automated renames of this variable will not update the name.  While this project is small, it doesn't help much, but as it grows it helps.

If there was a more meaningful message here, something along the lines of "Invalid FrameBuffer", it would warrant staying.  In general, less code is better since it speeds up reading.  This is also why static imports are helpful, since the word "Preconditions" doesn't add much meaning to the code.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 2 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Done.

File: core/src/main/java/io/grpc/transport/Http2ClientStream.java (hunk 4 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/Http2ClientStream.java]: Hmm, my hope was to move the common case to be the first thing in the method (thus switching the transport error check to look for null rather than not null).

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: You assume you know the exact version of gRPC being used. As it is today, just by looking at the backtrace I could figure out what is wrong, and not even need to know the version being run. That applies even if the name becomes outdated. Most issues that are filed with a backtrace don't include the version of the library that experienced the problem. As far as I'm concerned, for internal APIs the string could be random and still serve much of its purpose. For public APIs, having the string name match the argument is very helpful to users as it means they don't need to read the source.

I read `Precoditions.checkNotNull` as a single token; it doesn't take me any longer to read it than just `checkNotNull` and I don't have to assume that the method was statically imported (which is more applicable for `checkState`). I used to do static imports for Preconditions but stopped, because 1) it was causing frequent import noise (both additions and removals), especially with checkState, and 2) that import noise was causing merge conflicts. The banning of static wildcard imports is essentially the cause of my prefixing. I'm not against doing the same for `org.junit.Assert`, except the JUnit 3 history and the redundant "assert" in the method name.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java [hunk 5]
```diff
@@ -150,13 +155,14 @@ protected void inboundDataReceived(ReadableBuffer frame) {
...
   @Override
-  protected void inboundDeliveryPaused() {
+  protected final void inboundDeliveryPaused() {
     runCloseListenerTask();
   }
 
   @Override
   protected final void deframeFailed(Throwable cause) {
     log.log(Level.WARNING, "Exception processing message", cause);
+    // Should this be internal?
```
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: It probably should.

File: core/src/main/java/io/grpc/transport/Http2ClientStream.java (hunk 4 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/Http2ClientStream.java]: It seems less readable with the current change, so I'd just assume leave it as-is and respect the original author's style.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java [hunk 6]
```diff
@@ -45,9 +47,12 @@
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
+import javax.annotation.concurrent.NotThreadSafe;
+
 /**
  * The abstract base class for {@link ClientStream} implementations.
  */
+@NotThreadSafe
```
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: This is the very first usage of `NotThreadSafe`. We currently follow the standard practice of "something isn't thread-safe unless it says it is thread-safe." Are you thinking that we should start putting this on all classes, or this one is special for some reason?

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 1 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: So it sounds like there might be two things here.
1. There could be something that includes the version number in the stack trace  (something that catches it at a very high level, and tacks it on).  If knowing the version number is important, then a stray string in the message doesn't seem like a good way of propagating it.

2.  If the goal is to make it easy for users to see where they messed up, then a sentence fragment of what was expected would probably be more appropriate.  At least for internal APIs, the user may have missed providing some higher level parameter, which only caused indirectly (frame in this case) to be not set.  A full error message would be a better solution.

As for checkNotNull, A shorter name also can mean the difference between having to wrap a line and not.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 5 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Done.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 6 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Accidentally commented on the wrong PR, here's the snippet:

In regards to NotThreadSafe, Channel does use @ ThreadSafe, so there is at least some precedent. Additionally, since there is synchronization going on in AbstractStream, a new comer might be tempted to think that Streams are threadsafe.

Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: The @ ThreadSafe follows the practice that I described: "something isn't thread-safe unless it says it is thread-safe." It sounds like better commenting of AbstractStream is in order (maybe by using @ NotThreadSafe) instead of adding the annotation here.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: 1. My point was just that the exception, with the message, identifies the issue independent of gRPC version. There are many different places that exceptions can propagate out of the library; making sure all of them wrap the exception to include version information is 1) unreliable, 2) a major effort, and 3) dangerous, as it would likely change the type of exception.

2. I'm okay with using a sentence fragment instead. Since these `checkNotNull`s are lots of places, I do find it reasonable to use just the parameter name, since that is generally enough information. If certain places it seems necessary to add more information, I'm okay with that. But I don't want to increase the burden required for the string in all places because that is how you get people arguing to not include the string at all. Simply duplicating the parameter name is effective, easy, and is a pattern that can be reasonably be followed without deep thought.

I'm not worried about line-wrapping checkNotNull. I was using the full form on a project with 80 column wrapping and I don't recall it every being a problem. Do note that I'm not suggesting you to use the full form. I'm simply trying to argue that the original author gets to decide and we don't change it later without cause, as it is a style decision well within their right.

A strong percentage of merges I've had to do were regarding such import changes. Even if you aren't reviewing them, such line churn can create needless work, and I was tired of that needless work for the code I write. Again, it is a personal style decision, and the decision is left to the code author.

Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Let me add:
$ git grep " checkNotNull" | wc -l
13
$ git grep "\.checkNotNull" | wc -l
110

The current style is overwhelmingly to use the full form. You are free to use the short form in code you write, but it does seem unhelpful to change all usages in the current code.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java [hunk 7]
```diff
@@ -150,14 +155,14 @@ protected void inboundDataReceived(ReadableBuffer frame) {
...
-  protected void inboundDeliveryPaused() {
+  protected final void inboundDeliveryPaused() {
     runCloseListenerTask();
   }
 
   @Override
   protected final void deframeFailed(Throwable cause) {
     log.log(Level.WARNING, "Exception processing message", cause);
-    cancel(Status.CANCELLED);
+    cancel(Status.INTERNAL);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Hmm... actually, it looks like maybe we should remove the log and do:
cancel(Status.INTERNAL.withDescription("Exception processing message").withCause(cause));

This line is quite old and predates us having a clear idea on processing exceptions. Maybe leave it as you have (with INTERNAL), but have a TODO to investigate what is appropriate.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 1 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Added a more readable message.  In the future, if finding out the version the library being used is important, it should be outputted somewhere, rather than developers having to derive it from a stack trace message.  As it stands, the existing single variable name error messages are basically useless to any library users (as opposed to developers, who likely know exactly what version they are running anyways)

I am worried about line wrapping, after coming from a significantly larger project than this one, and having to line wrap a ton.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 3 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Added back a better message.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 6 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: Why not do both? Adding `NotThreadSafe` seems fine to me. The existence of the annotation doesn't automatically imply that it's inverting the default assumption about classes ... they'll still be assumed to be not thread safe unless specified otherwise.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 1 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: I'm in favor of leaving the `checkNotNull` as-is, but just because I feel the string is useful in and of itself, and that this is a convention used fairly widely (even within Google).

I do, however, disagree with the statement:

> the original author gets to decide and we don't change it later without cause, as it is a style decision well within their right.

This is open source code and the original author is irrelevant.  There is no code ownership and we do not need to consult the original author when tweaking code style.  The only requirement is that any submitted code adhere to the project coding conventions. Anything beyond that is fair game.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 6 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: I am more trying to understand why this class needs the annotation, specifically compared to other non-thread-safe clases. It seems we could put `NotThreadSafe` on every class that wasn't thread-safe, but to what end? I don't want to sign up for adding the annotation en masse. If there is a clear reason why this one is special, then I could be okay with adding it.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: @nmittler, I don't think you understand the intent of my quote. As it stands, if this PR went in as-is, I could make another PR that added back the `Preconditions`. Why wouldn't I do that, if _I_ like having the `Preconditions` more? It seems you don't have an argument as far as why that would be bad, but I do assert it _needs_ to be bad. For things where there are multiple reasonable options, we can't simply let it flip-flop between people's preferences. I felt that is dealt with by having some respect for the current state.

That isn't to say we couldn't change all the `checkNotNull`s. If we decided that not having `Preconditions` was better, then it would be fine to go and change the existing code. Similarly, if someone got bit by a `NullPointerException` that didn't have the message and took a lot of debugging to find, then adding the message could be reasonable.

Comment (nmittler) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: @ejona86 I guess I'm failing to see how changing a few methods to use static imports warrants a discussion at all.  Reducing the amount of code while maintaining understandability seems like a win-win to me. I think the likelihood of a ping pong between opposing developers is rather low ... certainly low enough that we don't have to be concerned by a static import.

Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: @nmittler, then after this PR gets checked in I'll send you a PR reverting the `checkNotNull` changes? I'll send out a cleanup PR that changes all usages of `checkNotNull` to be the non-static version.

I will note that we already have this issue today. This is why I keep on pointing on import order issues, because people have misconfigured IDEs and they keep fighting with each other. This is also why I point out other random reformatting changes that aren't necessary because I do know that people's IDEs are fighting for the formatting, and it does cause constant noise and merge conflicts, and people should fix their IDEs, deal with the annoyance of having to revert such changes, or stop using IDE features that are causing the trouble.

Comment (nmittler) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: @ejona86 IDE configuration issues are a slightly different matter and those sorts of things are violations of coding style and can (or at least should) be picked up by checkstyle. This topic, however, is a bit more squishy than that and is more in the realm of developer preference.

Yes, you could write a PR to change the style across the code base to your preference WRT `checkNotNull`, but why would you do that? It doesn't seem like that's quite the sort of thing that is being done here.  It's more of a "hey I think a few things in this file should change and while I'm in here I'll add a static import".  I see no malice in doing that.  And I see no problem with the use of a static import in this case as it does make the line more readable.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: @ejona86:  Java style guide says:
> You are advised not to comment on the following issues in code reviews:
> ...
> Import order. Don't itemize; just use the tools (see Section 9.3, Style-related tools).
> ...
> It is reasonable to ask the reviewee to check their editor configuration, but don't point out specific violations.

Code reviews are more of a sanity check, not an exact correctness check.  There is a list of [best practices](https://docs.google.com/document/d/1pG-WwM0cIWhH7SSuDXrAJ2JWX7JCYLI5HSRxMoTQPiY/), which includes

> Do not force your personal preferences. If you are suggesting to change something, there should be a clear win. If code complies with style guidelines and is not wrong, then code author is right.

It also stipulates that consistency in the code base should be enforced:
> Agree as a team to follow specific style where Java Style guide leaves it open-ended and enforce it

So, this means that the entire file should be changed at once, rather than just in single lines that changed.  It either needs to be up to the PR author, or consistently one way across the code, but not neither, and not both.   If you don't like the code review noise, then let's pick a hard-and-fast rule and follow it.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 7 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: @ejona86 Done

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 6 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: @ejona86: Documentation purposes.  It let's people know immediately this class is not suitable for thread safe use, without there being any doubt.

File: core/src/main/java/io/grpc/transport/Http2ClientStream.java (hunk 4 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/Http2ClientStream.java]: Done

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: > @ejona86: Java style guide says:
> 
> > You are advised not to comment on the following issues in code reviews:
> > ...
> > Import order. Don't itemize; just use the tools (see Section 9.3, Style-related tools).
> > ...
> > It is reasonable to ask the reviewee to check their editor configuration, but don't point out specific violations.

That is in the _internal_ style guide. We don't have the same tools. Internally we have linters that are very precise for imports and will automatically tell the reviewee of the wrong of their ways. Checkstyle is not as precise for imports. And this instances were causing problems. I think my argument still stands: a change messing up previously-correct formatting when not changing adjacent lines is harmful.

> Code reviews are more of a sanity check, not an exact correctness check. There is a list of best practices, which includes
> > Do not force your personal preferences. If you are suggesting to change something, there should be a clear win. If code complies with style guidelines and is not wrong, then code author is right.

This is fundamentally the same thing I am arguing. A _change_ needs to have a clear win. If it doesn't it shouldn't be made. I feel _you_ are forcing your personal preferences by changing existing code. If it was new code, I wouldn't have limited problem (I may still ask for a description to `checkNotNull()` for public interfaces).

I've not heard any reason why I couldn't make a PR immediately after this one that reverts back to the original form.

> It also stipulates that consistency in the code base should be enforced:
>> Agree as a team to follow specific style where Java Style guide leaves it open-ended and enforce it

> So, this means that the entire file should be changed at once, rather than just in single lines that changed. It either needs to be up to the PR author, or consistently one way across the code, but not neither, and not both. If you don't like the code review noise, then let's pick a hard-and-fast rule and follow it.

I don't comprehend the consistency argument used like you have. How is the removal of "Preconditions." in Http2ClientStream.java for "consistency"? It was consistent before your changes, and you did not add any Preconditions. You simply changed existing code that was already following the style guide to your personal preference.

How is that helpful? If that is permitted (which I assert it shouldn't be), then what would prevent us from just constantly changing it back and forth? What would prevent the use of full-file reformatting (which really can't be allowed)?

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 6 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: @carl-mastrangelo, I don't think you replied to my concern. How is this class different from any other non-thread-safe class (which is defined by "a class that did not specify it was thread-safe)? Why does _this_ class need it? I reject the idea of putting NotThreadSafe on every class, and so I want to make sure that it is clear that there should be a specific reason why it is added in each instance.

(I can even come up with some reasons myself, but I don't want to sign up for having the NotThreadSafes everywhere)

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: @ejona86  Alright.  I am going to revert this change since it doesn't look like this will be resolved.
